### PR TITLE
GPC-601: Handle no registrations

### PR DESCRIPTION
**Pull request by hannah-leong**

### CODE REVIEW

File: lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java [hunk 1]
```diff
@@ -133,6 +134,14 @@ class GroConvertToJsonTest {
...
+        "<RecordCount>0</RecordCount>" +
+        "</DeathRegistrationGroup>";
+    private static final String mockS3objectResponseExpectedRecordsButFoundNone =
+        "<DeathRegistrationGroup xmlns=\"http://www.ons.gov.uk/gro/OGDDeathExtractDWP\" xmlns:ns1=\"http://www.govtalk.gov.uk/people/PersonDescriptives\" xmlns:ns2=\"http://www.govtalk.gov.uk/people/AddressAndPersonalDetails\" xmlns:ns3=\"http://www.ons.gov.uk/gro/people/GROAddressDescriptives\" xmlns:ns4=\"http://www.ons.gov.uk/gro/people/GROPersonDescriptives\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://www.ons.gov.uk/gro/OGDDeathExtractDWP .\\OGDDeathExtractDWP-v1-1.xsd\">" +
+        "<RecordCount>6</RecordCount>" +
```
Comment (CarlyG55) [inline lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java:143]: The other example xml's don't seem to have a record count - is there a reason for this?

File: lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java [hunk 2]
```diff
@@ -107,4 +91,28 @@ private String getAuthorisationToken() {
...
         }
     }
+
+    private String convertXmlDataToJson(String xmlData, String authorisationToken) {
+        try {
+            var deathRegistrationGroup = xmlMapper.readValue(xmlData, DeathRegistrationGroup.class);
+            var records = deathRegistrationGroup.deathRegistrations();
+
+            if (records == null) {
+                if(deathRegistrationGroup.recordCount() == 0) {
```
Comment (CarlyG55) [inline lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java:101]: Do we have anywhere that checks the number of records is equal to the record count?

File: lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java (hunk 1 above)
Comment (hannah-leong) [reply lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java:143]: This is a mistake from previous testing as we did not check the values were equivalent and so did not need the RecordCount for the tests before. These will be added in when the checks for record count are implemented (next ticket).

File: lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java (hunk 2 above)
Comment (hannah-leong) [reply lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java:101]: This is listed as a separate ticket so will be on a different branch/in a different PR :) (will be ready v soon)

**Review by CarlyG55 — APPROVED** [review]

File: lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java [hunk 3]
```diff
@@ -107,4 +91,28 @@ private String getAuthorisationToken() {
...
+                    throw new MappingException("File contains no registration records");
```
Comment (olevett) [inline lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java:102]: I'm not entirely convinced this is an exception, thinking about it - it's plausibly just an empty file.

I'm happy enough keeping this though, and if it causes any problems/we get regular empty files then we can always change it later.

File: lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java [hunk 4]
```diff
@@ -133,6 +134,14 @@ class GroConvertToJsonTest {
         "</DeceasedBirthDate>" +
         "</DeathRegistration>" +
         "</DeathRegistrationGroup>";
+    private static final String mockS3objectResponseNoRecords =
```
Comment (olevett) [inline lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java:137]: Could we use the """ format for thi?

File: lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java [hunk 5]
```diff
@@ -133,6 +134,14 @@ class GroConvertToJsonTest {
...
+        "<DeathRegistrationGroup xmlns=\"http://www.ons.gov.uk/gro/OGDDeathExtractDWP\" xmlns:ns1=\"http://www.govtalk.gov.uk/people/PersonDescriptives\" xmlns:ns2=\"http://www.govtalk.gov.uk/people/AddressAndPersonalDetails\" xmlns:ns3=\"http://www.ons.gov.uk/gro/people/GROAddressDescriptives\" xmlns:ns4=\"http://www.ons.gov.uk/gro/people/GROPersonDescriptives\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://www.ons.gov.uk/gro/OGDDeathExtractDWP .\\OGDDeathExtractDWP-v1-1.xsd\">" +
+        "<RecordCount>0</RecordCount>" +
+        "</DeathRegistrationGroup>";
```
Comment (olevett) [inline lambdas/gro-convert-to-json/src/test/java/uk/gov/di/data/lep/GroConvertToJsonTest.java:140]:
```suggestion
    private static final String mockS3objectResponseOneRecord =
            """
            <DeathRegistrationGroup
                xmlns="http://www.ons.gov.uk/gro/OGDDeathExtractDWP"
                xmlns:ns1="http://www.govtalk.gov.uk/people/PersonDescriptives"
                xmlns:ns2="http://www.govtalk.gov.uk/people/AddressAndPersonalDetails"
                xmlns:ns3="http://www.ons.gov.uk/gro/people/GROAddressDescriptives" xmlns:ns4="http://www.ons.gov.uk/gro/people/GROPersonDescriptives"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                xsi:schemaLocation="http://www.ons.gov.uk/gro/OGDDeathExtractDWP .\\OGDDeathExtractDWP-v1-1.xsd">
                <DeathRegistration>
                    <RegistrationID>1</RegistrationID>
                    <DeceasedName>
                        <ns4:PersonGivenName>ERICA</ns4:PersonGivenName>
                        <ns4:PersonGivenName>CHRISTINA</ns4:PersonGivenName>
                        <ns4:PersonFamilyName>BLOGG</ns4:PersonFamilyName>
                    </DeceasedName>
                    <DeceasedGender>2</DeceasedGender>
                    <DeceasedAddress>
                        <ns3:Line>263 Ave Maria Lane</ns3:Line>
                        <ns3:Line>Bourton-on-the-hill</ns3:Line>
                        <ns3:Postcode>BT62 4HL</ns3:Postcode>
                    </DeceasedAddress>
                </DeathRegistration>
            </DeathRegistrationGroup>
            """;
```

File: lib/src/main/java/uk/gov/di/data/lep/library/exceptions/MappingException.java [hunk 6]
```diff
@@ -4,4 +4,7 @@ public class MappingException extends RuntimeException {
     public MappingException(Exception e) {
         super(e);
     }
+    public MappingException(String message) {
```
Comment (olevett) [inline lib/src/main/java/uk/gov/di/data/lep/library/exceptions/MappingException.java:7]:
```suggestion

    public MappingException(String message) {
```

File: lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java (hunk 2 above)
Comment (olevett) [inline lambdas/gro-convert-to-json/src/main/java/uk/gov/di/data/lep/GroConvertToJson.java:101]:
```suggestion
                if (deathRegistrationGroup.recordCount() == 0) {
```

**Pull request merged**